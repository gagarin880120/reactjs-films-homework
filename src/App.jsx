import { hot } from 'react-hot-loader/root';
import React from 'react';
import MovieDetailsPage from './pages/MovieDetailsPage/MovieDetailsPage';

function App() {
  return (
    <>
      <MovieDetailsPage />
    </>
  );
}

export default hot(App);
