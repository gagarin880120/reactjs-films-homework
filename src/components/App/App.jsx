import { hot } from 'react-hot-loader/root';
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import {
  Switch,
  Route,
  Redirect,
  useLocation,
} from 'react-router-dom';
import MovieDetailsPageContainer from '../../pages/MovieDetailsPage';
import SearchResultsPageContainer from '../../pages/SearchResultsPage';
import styles from './App.module.scss';
import Spinner from '../Spinner/Spinner';

function App({
  onAppLoad, genres, onReload, currentAPIRequest,
}) {
  const location = useLocation();
  useEffect(() => {
    if (!genres) {
      onAppLoad();
    }
    if (location.pathname.length > 1) {
      onReload(location.pathname);
    }
  }, [location]);

  return (
    genres
      ? (
        <div className={styles.wrapper}>
          <Switch>
            <Route exact path="/">
              <Redirect to={`/${currentAPIRequest}`} />
            </Route>
            <Route path={`/${currentAPIRequest}`} exact component={SearchResultsPageContainer} />
            <Route path="/movie/:id" exact component={MovieDetailsPageContainer} />
          </Switch>
        </div>
      ) : (
        <div className={styles.loading}>
          <Spinner />
        </div>
      )
  );
}

export default hot(App);

App.propTypes = {
  onAppLoad: PropTypes.func,
  genres: PropTypes.instanceOf(Array),
  currentAPIRequest: PropTypes.string,
  onReload: PropTypes.func,
};

App.defaultProps = {
  onAppLoad: null,
  genres: [],
  currentAPIRequest: '',
  onReload: null,
};
