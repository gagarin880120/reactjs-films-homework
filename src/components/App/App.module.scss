.wrapper {
  font-family: Arial, Helvetica, sans-serif;
  color: #fff;
}

.loading {
  margin-top: 20%;
}
