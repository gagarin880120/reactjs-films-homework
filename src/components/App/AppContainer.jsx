import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  getGenres, setCurrentAPIRequest, getMovieDetails, getMovies,
} from '../../redux/actions';
import { genresSelector, currentAPIRequestSelector } from '../../redux/selectors';
import App from './App';

export function AppContainer({
  onAppLoad, genres, currentAPIRequest, onReload,
}) {
  return (
    <Router>
      <App
        onAppLoad={onAppLoad}
        genres={genres}
        currentAPIRequest={currentAPIRequest}
        onReload={onReload}
      />
    </Router>
  );
}

export const mapStateToProps = (state) => ({
  genres: genresSelector(state),
  currentAPIRequest: currentAPIRequestSelector(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onAppLoad() {
    dispatch(getGenres());
  },
  onReload(path) {
    if (path.includes('movie/')) {
      dispatch(getMovieDetails(path.slice(path.lastIndexOf('/') + 1)));
      dispatch(getMovies('popular', 1));
    } else {
      dispatch(setCurrentAPIRequest(path.slice(path.indexOf('/') + 1)));
    }
  },

});

AppContainer.propTypes = {
  onAppLoad: PropTypes.func,
  genres: PropTypes.instanceOf(Array),
  currentAPIRequest: PropTypes.string,
  onReload: PropTypes.func,
};

AppContainer.defaultProps = {
  onAppLoad: null,
  genres: [],
  currentAPIRequest: '',
  onReload: null,
};

export default connect(mapStateToProps, mapDispatchToProps)(AppContainer);
