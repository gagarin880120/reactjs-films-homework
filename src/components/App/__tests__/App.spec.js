import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import App from '../App';
import { create, act } from 'react-test-renderer';
import { useLocation } from 'react-router-dom';

jest.mock('react-router-dom', () => ({
  useLocation: jest.fn().mockReturnValue({pathname: '/popular'}),
  Switch: () => null,
  Route: () => null,
  Redirect: jest.fn(),
}));

describe('App component', () => {
  test('should render without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<App />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  test('should call onAppLoad when is mounted and are not genres', () => {
    let root;
    const onAppLoad = jest.fn();
    const onReload = jest.fn();

    act(() => {
      root = create(
        <App onAppLoad={onAppLoad} onReload={onReload} genres={null} />
      )
    });

    expect(onAppLoad).toHaveBeenCalled();
    expect(onReload).toHaveBeenCalled();
  });

  test('should not call onAppLoad when is mounted and are genres', () => {
    const onAppLoad = jest.fn();
    const onReload = jest.fn();
    useLocation.mockReturnValue({pathname: '/'});
    let root;

    act(() => {
      root = create(
        <App onAppLoad={onAppLoad} onReload={onReload} genres={[{name: 'Adventure', id: 28}]} />
      )
    });

    expect(onAppLoad).not.toHaveBeenCalled();
    expect(onReload).not.toHaveBeenCalled();
  });
});
