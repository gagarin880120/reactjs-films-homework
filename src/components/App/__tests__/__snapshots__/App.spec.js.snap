// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render without crashing 1`] = `
<div
  className="wrapper"
>
  <Switch>
    <Route
      exact={true}
      path="/"
    >
      <mockConstructor
        to="/"
      />
    </Route>
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(SearchResultsPageContainer)",
          "type": [Function],
        }
      }
      exact={true}
      path="/"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(MovieDetailsPageContainer)",
          "type": [Function],
        }
      }
      exact={true}
      path="/movie/:id"
    />
  </Switch>
</div>
`;
